src/shared/bootspec.c

```diff
@@ -529,10 +531,17 @@ static int boot_entry_compare(const BootEntry *a, const BootEntry *b) {
                         return r;
         }
 
-        r = -strverscmp_improved(a->id, b->id);
+        r = -strverscmp_improved(a->id_without_profile ?: a->id, b->id_without_profile ?: b->id);
         if (r != 0)
                 return r;
 
+        if (a->id_without_profile && b->id_without_profile &&
+            streq(a->id_without_profile, b->id_without_profile)) {
```

Comment:
The same comment as https://github.com/systemd/systemd/pull/34339#discussion_r1751718593.
`streq()` can be dropped.